Source/DafnyCore/Resolver/PreType/PreType.cs

```diff
@@ -0,0 +1,326 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public abstract class PreType {
+    public PreType Normalize() {
+      var t = this;
+      while (t is PreTypeProxy proxy && proxy.PT != null) {
+        t = proxy.PT;
+      }
+      return t;
+    }
+
+    public static bool Same(PreType a, PreType b) {
+      a = a.Normalize();
+      b = b.Normalize();
+      if (a is DPreType ap && b is DPreType bp && ap.Decl == bp.Decl) {
+        Contract.Assert(ap.Arguments.Count == bp.Arguments.Count);
```

Comment:
This inner part could be moved into a separate `Same` function in `DPreType`.

Reply:
It's such a small piece of code that I'd prefer to keep it here. But I did replace the loop with one `... Zip ... All ...` construction.